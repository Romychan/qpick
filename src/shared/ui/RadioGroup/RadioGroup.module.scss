.container {
  display: flex;

  &.horizontal {
    flex-direction: row;
    gap: 1rem;
  }

  &.vertical {
    flex-direction: column;
    gap: 0.25rem;
  }
}
