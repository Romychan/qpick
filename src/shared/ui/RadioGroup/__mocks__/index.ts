export const MOCK_RADIO_ITEMS = [
  { label: 'Item #1', value: '1' },
  { label: 'Item #2', value: '2' },
];
