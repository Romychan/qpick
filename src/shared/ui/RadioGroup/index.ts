export { RadioGroup } from './RadioGroup';
export type { RadioOptionType } from './types';
