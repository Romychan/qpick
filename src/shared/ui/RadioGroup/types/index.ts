/** Base radio item interface */
export interface RadioOptionType<T = string> {
  /** The displayed value of the element */
  label: string;
  /** The unique value of the element */
  value: T;
}
