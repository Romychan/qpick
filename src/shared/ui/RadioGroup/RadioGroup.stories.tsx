import type { Meta, StoryObj } from '@storybook/react';
import { useArgs } from '@storybook/preview-api';
import { action } from '@storybook/addon-actions';

import { withGapDecorator, withWidthDecorator } from '../../lib/storybook';

import { RadioGroup } from './RadioGroup';
import { MOCK_RADIO_ITEMS } from './__mocks__';

const meta: Meta<typeof RadioGroup> = {
  title: 'UI/Radio/RadioGroup',
  component: RadioGroup,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    options: MOCK_RADIO_ITEMS,
  },
};

export default meta;
type Story = StoryObj<typeof RadioGroup>;

/** This state is used to default. */
export const Default: Story = {
  decorators: [withWidthDecorator],
  args: {
    selectedValue: '1',
  },
  render: function Render(args) {
    const [{ selectedValue }, updateArgs] = useArgs();

    const handleChange = (value: string) => {
      action('onChange')(value);
      updateArgs({ selectedValue: value });
    };

    return (
      <RadioGroup
        {...args}
        selectedValue={selectedValue}
        onChange={handleChange}
      />
    );
  },
};

/**
 * The `direction` property is used to change the direction of positions radio items.
 * - `horizontal` is used to direction radio items horizontally
 * - `vertical` is used to direction radio items vertically
 */
export const Variants: Story = {
  decorators: [withGapDecorator],
  render: (args) => {
    return (
      <>
        <RadioGroup {...args} direction="horizontal" />
        <RadioGroup {...args} direction="vertical" />
      </>
    );
  },
};
