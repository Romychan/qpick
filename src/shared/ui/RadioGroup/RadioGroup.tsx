import { cl } from '../../lib/utils';
import { Radio } from '../Radio/Radio';

import { RadioOptionType } from './types';
import styles from './RadioGroup.module.scss';

interface IRadioGroupProps {
  /** List of options to display */
  options: RadioOptionType[];
  /** The current selected value */
  selectedValue?: string;
  /** Additional class for the component */
  className?: string;
  /**
   * The direction of the radio items
   *
   * @default 'horizontal'
   */
  direction?: 'horizontal' | 'vertical';
  /**
   * The callback that will be called when the radio item is selected
   *
   * @param value Selected value
   */
  onChange: (value: string) => void;
}

/** This component is used to select one option from a list of all options */
export const RadioGroup = ({
  options,
  selectedValue,
  direction = 'horizontal',
  className,
  onChange,
}: IRadioGroupProps) => {
  return (
    <div
      className={cl(styles.container, styles[direction], className)}
      data-testid="radio-group"
    >
      {options.map(({ value, label }) => (
        <Radio
          key={value}
          id={value}
          name={value}
          value={value}
          label={label}
          checked={value === selectedValue}
          onChange={(event) => onChange(event.target.value)}
        />
      ))}
    </div>
  );
};
