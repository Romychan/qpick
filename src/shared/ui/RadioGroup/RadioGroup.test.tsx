import { fireEvent, render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import { RadioGroup } from './RadioGroup';
import { MOCK_RADIO_ITEMS } from './__mocks__';

describe('RadioGroup', () => {
  it('should render correctly', () => {
    const handleChange = vi.fn();
    render(<RadioGroup options={MOCK_RADIO_ITEMS} onChange={handleChange} />);

    const groupElement = screen.getByTestId('radio-group');
    const itemsElement = screen.getAllByRole('radio');

    expect(groupElement).toBeInTheDocument();
    expect(itemsElement).toHaveLength(2);
  });

  it('should render correctly with direction', () => {
    const handleChange = vi.fn();
    render(
      <RadioGroup
        direction="horizontal"
        options={MOCK_RADIO_ITEMS}
        onChange={handleChange}
      />,
    );

    const groupElement = screen.getByTestId('radio-group');

    expect(groupElement.getAttribute('class')).toMatch('horizontal');
    expect(groupElement.getAttribute('class')).not.toMatch('vertical');
  });

  it('should return the selected option correctly', () => {
    const handleChange = vi.fn();
    render(<RadioGroup options={MOCK_RADIO_ITEMS} onChange={handleChange} />);

    const itemElement = screen.getByRole('radio', { name: 'Item #1' });

    fireEvent.click(itemElement);

    expect(handleChange).toHaveBeenCalledWith('1');
  });

  it('should match snapshot', () => {
    const handleChange = vi.fn();
    const container = render(
      <RadioGroup options={MOCK_RADIO_ITEMS} onChange={handleChange} />,
    );

    expect(container).toMatchSnapshot();
  });
});
