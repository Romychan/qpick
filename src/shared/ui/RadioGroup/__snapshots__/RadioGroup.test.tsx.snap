// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioGroup > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_00874a _horizontal_00874a"
        data-testid="radio-group"
      >
        <label
          class="_container_d7e7cf"
          for="1"
        >
          <input
            class="_input_d7e7cf"
            id="1"
            name="1"
            type="radio"
            value="1"
          />
          <span
            class="_check_d7e7cf"
          />
          <span
            class="_label_d7e7cf"
          >
            Item #1
          </span>
        </label>
        <label
          class="_container_d7e7cf"
          for="2"
        >
          <input
            class="_input_d7e7cf"
            id="2"
            name="2"
            type="radio"
            value="2"
          />
          <span
            class="_check_d7e7cf"
          />
          <span
            class="_label_d7e7cf"
          >
            Item #2
          </span>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_00874a _horizontal_00874a"
      data-testid="radio-group"
    >
      <label
        class="_container_d7e7cf"
        for="1"
      >
        <input
          class="_input_d7e7cf"
          id="1"
          name="1"
          type="radio"
          value="1"
        />
        <span
          class="_check_d7e7cf"
        />
        <span
          class="_label_d7e7cf"
        >
          Item #1
        </span>
      </label>
      <label
        class="_container_d7e7cf"
        for="2"
      >
        <input
          class="_input_d7e7cf"
          id="2"
          name="2"
          type="radio"
          value="2"
        />
        <span
          class="_check_d7e7cf"
        />
        <span
          class="_label_d7e7cf"
        >
          Item #2
        </span>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
